File: backend/src/main/java/dev/frilly/locket/controller/RequestsController.java
package dev.frilly.locket.controller;

import dev.frilly.locket.data.FriendRequest;
import dev.frilly.locket.data.Friendship;
import dev.frilly.locket.data.User;
import dev.frilly.locket.dto.req.UsernameRequest;
import dev.frilly.locket.dto.res.PaginatedResponse;
import dev.frilly.locket.dto.res.UserResponse;
import dev.frilly.locket.repo.FriendRequestRepository;
import dev.frilly.locket.repo.FriendshipRepository;
import dev.frilly.locket.repo.UserRepository;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.http.HttpStatus;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.server.ResponseStatusException;

import java.util.function.Function;

/**
 * Controller for handling friend requests.
 */
@RestController
public final class RequestsController {

  @Autowired
  private FriendRequestRepository frRepo;

  @Autowired
  private UserRepository userRepo;

  @Autowired
  private FriendshipRepository fsRepo;

  /**
   * GET /requests.
   * <p>
   * Retrieves a list of my friend requests.
   */
  @GetMapping("/requests")
  public PaginatedResponse<UserResponse> getRequests(
      @RequestParam(value = "page", defaultValue = "0") int page,
      @RequestParam(value = "per_page", defaultValue = "20") int perPage,
      @RequestParam(value = "myself", defaultValue = "false") boolean myself
  ) {
    final var auth = SecurityContextHolder.getContext().getAuthentication();
    final var user = (User) auth.getPrincipal();

    final Page<FriendRequest>                   query;
    final Function<FriendRequest, UserResponse> mapFunction;

    if (myself) {
      query       = frRepo.findAllBySender(user,
          PageRequest.of(page, perPage, Sort.by("receiver.id")));
      mapFunction = fr -> fr.receiver().makeResponse();
    } else {
      query       = frRepo.findAllByReceiver(user,
          PageRequest.of(page, perPage, Sort.by("sender.id")));
      mapFunction = fr -> fr.sender().makeResponse();
    }

    return new PaginatedResponse<>(query.getTotalElements(),
        query.getTotalPages(), page, perPage,
        query.get().map(mapFunction).toList());
  }

  /**
   * POST /requests.
   * Sends a new request.
   */
  @PostMapping("/requests")
  public UserResponse postRequests(@RequestBody final UsernameRequest body) {
    final var auth   = SecurityContextHolder.getContext().getAuthentication();
    final var user   = (User) auth.getPrincipal();
    final var target = userRepo.findByUsername(body.username());

    if (target.isEmpty()) {
      throw new ResponseStatusException(HttpStatus.NOT_FOUND);
    }

    if (user.id() == target.get().id()) {
      throw new ResponseStatusException(HttpStatus.FORBIDDEN);
    }

    if (frRepo.findBySenderAndReceiver(user, target.get()).isPresent()) {
      throw new ResponseStatusException(HttpStatus.CONFLICT);
    }

    final var req = frRepo.findBySenderAndReceiver(target.get(), user);
    if (req.isPresent()) {
      frRepo.delete(req.get());

      final Friendship fs;
      if (user.id() < target.get().id()) {
        fs = new Friendship(user, target.get());
      } else {
        fs = new Friendship(target.get(), user);
      }

      fsRepo.save(fs);
      throw new ResponseStatusException(HttpStatus.NO_CONTENT);
    }

    final var fr = new FriendRequest(user, target.get());
    frRepo.save(fr);
    return target.get().makeResponse();
  }

  /**
   * DELETE /requests
   * Delete a friend request sent.
   */
  @DeleteMapping("/requests")
  public UserResponse deleteRequest(@Valid @RequestBody final UsernameRequest body) {
    final var auth   = SecurityContextHolder.getContext().getAuthentication();
    final var user   = (User) auth.getPrincipal();
    final var target = userRepo.findByUsername(body.username());

    if (target.isEmpty()) {
      throw new ResponseStatusException(HttpStatus.NOT_FOUND);
    }

    final var sentFr     = frRepo.findBySenderAndReceiver(user, target.get());
    final var receivedFr = frRepo.findBySenderAndReceiver(target.get(), user);

    if (sentFr.isEmpty() && receivedFr.isEmpty()) {
      throw new ResponseStatusException(HttpStatus.NO_CONTENT);
    }

    sentFr.ifPresent(frRepo::delete);
    receivedFr.ifPresent(frRepo::delete);
    return target.get().makeResponse();
  }

}

File: backend/src/main/kotlin/dev/frilly/locket/controller/RequestsController.kt
package dev.frilly.locket.controller

import dev.frilly.locket.controller.dto.AbstractUser
import dev.frilly.locket.controller.dto.GetRequestsResponse
import dev.frilly.locket.controller.dto.PostRequestsRequest
import dev.frilly.locket.data.FriendRequest
import dev.frilly.locket.data.Friendship
import dev.frilly.locket.data.FriendshipCompositeKey
import dev.frilly.locket.data.User
import dev.frilly.locket.repo.FriendRequestRepository
import dev.frilly.locket.repo.FriendshipRepository
import dev.frilly.locket.repo.UserRepository
import jakarta.validation.Valid
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.web.bind.annotation.*
import kotlin.jvm.optionals.getOrNull

/**
 * Controller handler for friend requests (/requests).
 */
@RestController
class RequestsController {

    @Autowired
    private lateinit var requestRepository: FriendRequestRepository

    @Autowired
    private lateinit var userRepository: UserRepository

    @Autowired
    private lateinit var friendshipRepository: FriendshipRepository

    /**
     * GET /requests.
     * Retrieves a list of my friend requests.
     */
    @GetMapping("/requests")
    fun getRequests(): GetRequestsResponse {
        val auth = SecurityContextHolder.getContext().authentication
        val user = auth.principal as User

        return GetRequestsResponse(
            results = requestRepository.findAllByReceiver(user).map {
                AbstractUser(
                    it.sender.id,
                    it.sender.username,
                    it.sender.avatarUrl,
                )
            }
        )
    }

    /**
     * POST /requests.
     * Sends a new request.
     */
    @PostMapping("/requests")
    fun postRequests(@Valid @RequestBody body: PostRequestsRequest)
            : ResponseEntity<PostRequestsRequest> {
        val auth = SecurityContextHolder.getContext().authentication
        val user = auth.principal as User
        val target = userRepository.findByUsername(body.username).getOrNull()
            ?: return ResponseEntity.notFound().build()

        if (user.id == target.id) {
            return ResponseEntity.status(HttpStatus.FORBIDDEN).build()
        }

        if (requestRepository.findBySenderAndReceiver(user, target).isPresent) {
            return ResponseEntity(HttpStatus.CONFLICT)
        }

        // If the other already sent a request, we can now "accept" it.
        val req = requestRepository.findBySenderAndReceiver(target, user)
        if (req.isPresent) {
            requestRepository.delete(req.get())

            val friendship =
                if (user.id < target.id)
                    Friendship(
                        id = FriendshipCompositeKey(user.id, target.id),
                        user1 = user,
                        user2 = target,
                    )
                else
                    Friendship(
                        id = FriendshipCompositeKey(target.id, user.id),
                        user1 = target,
                        user2 = user,
                    )
            friendshipRepository.save(friendship)
            return ResponseEntity.noContent().build()
        }

        val obj = FriendRequest(
            id = 0,
            sender = user,
            receiver = target,
        )
        requestRepository.save(obj)
        return ResponseEntity.ok(PostRequestsRequest(target.username))
    }

    /**
     * DELETE /requests
     * Delete a friend request sent.
     */
    @DeleteMapping("/requests")
    fun deleteRequests(@Valid @RequestBody body: PostRequestsRequest)
            : ResponseEntity<PostRequestsRequest> {
        val auth = SecurityContextHolder.getContext().authentication
        val user = auth.principal as User
        val target = userRepository.findByUsername(body.username).getOrNull()
            ?: return ResponseEntity.notFound().build()

        val obj =
            requestRepository.findBySenderAndReceiver(user, target).getOrNull()
                ?: return ResponseEntity.noContent().build()
        requestRepository.delete(obj)
        return ResponseEntity.ok(PostRequestsRequest(target.username))
    }

}
